program/src/processor.rs

```diff
@@ -0,0 +1,485 @@
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    // [Core BPF]: This check _is required_ since
+    // "FKAcEvNgSY79RpqsPNUV5gDyumopH4cEHqUxyfm8b8Ap" was activated on
+    // mainnet-beta.
+    // See https://github.com/solana-labs/solana/blob/e4064023bf7936ced97b0d4de22137742324983d/programs/address-lookup-table/src/processor.rs#L129-L135.
+    if check_id(lookup_table_info.owner) {
+        return Ok(());
+    }
+
+    let lookup_table_data_len = LOOKUP_TABLE_META_SIZE;
+    let rent = <Rent as Sysvar>::get()?;
+    let required_lamports = rent
+        .minimum_balance(lookup_table_data_len)
+        .max(1)
+        .saturating_sub(lookup_table_info.lamports());
+
+    if required_lamports > 0 {
+        invoke(
+            &system_instruction::transfer(payer_info.key, lookup_table_info.key, required_lamports),
+            &[payer_info.clone(), lookup_table_info.clone()],
+        )?;
+    }
+
+    invoke_signed(
+        &system_instruction::allocate(lookup_table_info.key, lookup_table_data_len as u64),
+        &[lookup_table_info.clone()],
+        &[&[
+            authority_info.key.as_ref(),
+            &derivation_slot.to_le_bytes(),
+            &[bump_seed],
+        ]],
+    )?;
+
+    invoke_signed(
+        &system_instruction::assign(lookup_table_info.key, program_id),
+        &[lookup_table_info.clone()],
+        &[&[
+            authority_info.key.as_ref(),
+            &derivation_slot.to_le_bytes(),
+            &[bump_seed],
+        ]],
+    )?;
+
+    ProgramState::serialize_new_lookup_table(
+        *lookup_table_info.try_borrow_mut_data()?,
+        authority_info.key,
+    )?;
+
+    Ok(())
+}
+
+fn process_freeze_lookup_table(program_id: &Pubkey, accounts: &[AccountInfo]) -> ProgramResult {
+    let accounts_iter = &mut accounts.iter();
+
+    let lookup_table_info = next_account_info(accounts_iter)?;
+    let authority_info = next_account_info(accounts_iter)?;
+
+    if lookup_table_info.owner != program_id {
+        return Err(ProgramError::InvalidAccountOwner);
+    }
+
+    if !authority_info.is_signer {
+        msg!("Authority account must be a signer");
+        return Err(ProgramError::MissingRequiredSignature);
+    }
+
+    let mut lookup_table_meta = {
+        let lookup_table_data = lookup_table_info.try_borrow_data()?;
+        let lookup_table = AddressLookupTable::deserialize(&lookup_table_data)?;
+
+        if lookup_table.meta.authority.is_none() {
+            msg!("Lookup table is already frozen");
+            // [Core BPF]: TODO: Should be `ProgramError::Immutable`
+            // See https://github.com/solana-labs/solana/pull/35113
+            return Err(ProgramError::Custom(0));
+        }
+        if lookup_table.meta.authority != Some(*authority_info.key) {
+            // [Core BPF]: TODO: Should be `ProgramError::IncorrectAuthority`
+            // See https://github.com/solana-labs/solana/pull/35113
+            return Err(ProgramError::Custom(0));
+        }
```

Comment:
nit: how about adding a `msg!` like with the other error cases?

Reply:
Sure! Left out intentionally, since the original did not. I'll add, as well as the other ones you suggested.